File: skill/plugin/reranker.ts
/**
 * TotalReclaw Plugin -- Reranker (thin wrapper around `@totalreclaw/core`).
 *
 * As of rc.22 the plugin no longer ships its own BM25 / RRF / source-weight
 * implementation. All ranking decisions are delegated to the canonical Rust
 * core (`rerankWithConfig`, `cosineSimilarity`, `sourceWeight`,
 * `legacyClaimFallbackWeight`) so plugin / Hermes / MCP runtimes share one
 * source of truth and cannot drift again (rc.18 cosine-gate divergence).
 *
 * The previous client-side pipeline added importance, recency, and MMR
 * signals on top of BM25 + cosine. Those signals are dropped here:
 * core's intent-weighted RRF + source-weighted final score is the
 * canonical Pipeline G + Tier 1 mix that benchmark E13 calibrated. The
 * extra TS-side passes were not part of the validated baseline and were a
 * source of cross-client divergence.
 *
 * Public surface kept for callers (index.ts, semantic-dedup.ts,
 * consolidation.ts, pocv2-e2e-test.ts, v1-taxonomy.test.ts):
 *   - `rerank(query, queryEmbedding, candidates, topK, _legacyWeights, applySourceWeights)`
 *   - `cosineSimilarity(a, b)`
 *   - `getSourceWeight(source)`
 *   - `detectQueryIntent(query)` and `INTENT_WEIGHTS`
 *     (kept as no-ops for callers that pass them in -- core handles
 *      intent-weighting internally now, so the legacy weights argument is
 *      ignored.)
 *   - Types: `RerankerCandidate`, `RerankerResult`, `RankingWeights`,
 *     `QueryIntent`.
 */

import * as core from '@totalreclaw/core';

// ---------------------------------------------------------------------------
// Cosine Similarity (delegated to core WASM)
// ---------------------------------------------------------------------------

export function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length === 0 || b.length === 0) return 0;
  if (a.length !== b.length) return 0;
  return core.cosineSimilarity(new Float32Array(a), new Float32Array(b));
}

// ---------------------------------------------------------------------------
// Source-weight lookup (delegated to core)
// ---------------------------------------------------------------------------

export function getSourceWeight(source: string | undefined): number {
  if (!source) return core.legacyClaimFallbackWeight();
  return core.sourceWeight(source);
}

// ---------------------------------------------------------------------------
// Query Intent Detection
//
// Kept as a no-op compat layer so existing call sites that still pass
// INTENT_WEIGHTS[intent] into `rerank()` continue to compile. Core does its
// own intent-weighting internally based on the per-candidate cosine score
// (see `rust/totalreclaw-core/src/reranker.rs` -- `bm25_weight = 0.3 + 0.3 *
// (1 - intent_score)`) so the TS-side weight argument is ignored.
// ---------------------------------------------------------------------------

export type QueryIntent = 'factual' | 'temporal' | 'semantic';

export interface RankingWeights {
  bm25: number;
  cosine: number;
  importance: number;
  recency: number;
}

const TEMPORAL_KEYWORDS = /\b(yesterday|today|last\s+week|last\s+month|recently|recent|latest|ago|when|this\s+week|this\s+month|earlier|before|after|since|during|tonight|morning|afternoon)\b/i;
const FACTUAL_PATTERNS = /^(what|who|where|which|how\s+many|how\s+much|is\s+|are\s+|does\s+|do\s+|did\s+|was\s+|were\s+)\b/i;

export const INTENT_WEIGHTS: Record<QueryIntent, RankingWeights> = {
  factual:  { bm25: 0.40, cosine: 0.20, importance: 0.25, recency: 0.15 },
  temporal: { bm25: 0.15, cosine: 0.20, importance: 0.20, recency: 0.45 },
  semantic: { bm25: 0.20, cosine: 0.35, importance: 0.25, recency: 0.20 },
};

export function detectQueryIntent(query: string): QueryIntent {
  if (TEMPORAL_KEYWORDS.test(query)) return 'temporal';
  if (FACTUAL_PATTERNS.test(query) && query.length < 80) return 'factual';
  return 'semantic';
}

// ---------------------------------------------------------------------------
// Candidate / Result types (cross-runtime stable)
// ---------------------------------------------------------------------------

export interface RerankerCandidate {
  id: string;
  text: string;
  embedding?: number[];
  /** Unused now -- core ignores importance. Kept for API stability. */
  importance?: number;
  /** Unused now -- core ignores recency. Kept for API stability. */
  createdAt?: number;
  /** Memory Taxonomy v1 source ("user" | "user-inferred" | ... ). */
  source?: string;
}

export interface RerankerResult extends RerankerCandidate {
  /** Final fused score from core (post source-weight multiplication). */
  rrfScore: number;
  cosineSimilarity?: number;
  /** Source weight multiplier applied (1.0 if no weighting). */
  sourceWeight?: number;
}

// Core's RankedResult JSON shape (returned by rerankWithConfig).
interface CoreRankedResult {
  id: string;
  text: string;
  score: number;
  bm25_score: number;
  cosine_score: number;
  timestamp: string;
  source_weight?: number;
}

// ---------------------------------------------------------------------------
// Re-ranker (delegates to core::reranker)
// ---------------------------------------------------------------------------

/**
 * Re-rank decrypted candidates by delegating to core's `rerankWithConfig`.
 *
 * @param query              The user's plaintext search query.
 * @param queryEmbedding     Embedding vector for the query.
 * @param candidates         Decrypted candidates with text and optional
 *                           embeddings + v1 source.
 * @param topK               Top-K to return (default 8).
 * @param _legacyWeights     IGNORED -- kept for API stability so existing
 *                           callers still compile. Core handles
 *                           intent-weighting internally based on per-candidate
 *                           cosine scores.
 * @param applySourceWeights When true, multiply the final fused score by
 *                           the v1 source-weight (Retrieval v2 Tier 1).
 *                           Default true at all production call sites.
 */
export function rerank(
  query: string,
  queryEmbedding: number[],
  candidates: RerankerCandidate[],
  topK: number = 8,
  _legacyWeights?: Partial<RankingWeights>,
  applySourceWeights: boolean = false,
): RerankerResult[] {
  if (candidates.length === 0) return [];

  // Build the core::Candidate JSON shape. Core requires:
  //   { id, text, embedding: number[], timestamp: string, source? }
  // Candidates without an embedding pass an empty array; core's cosine
  // similarity returns 0 for that case (see cosine_similarity_f32).
  const coreCandidates = candidates.map((c) => ({
    id: c.id,
    text: c.text,
    embedding: c.embedding ?? [],
    timestamp: c.createdAt != null ? String(c.createdAt) : '',
    ...(c.source ? { source: c.source } : {}),
  }));

  // Empty queryEmbedding is allowed: core will compute cosine = 0 across
  // all candidates and rely entirely on BM25 (intent_score clamps to 0
  // -> bm25_weight = 0.6, cosine_weight = 0.3).
  const queryVec = new Float32Array(queryEmbedding ?? []);
  const candidatesJson = JSON.stringify(coreCandidates);

  const ranked = core.rerankWithConfig(
    query,
    queryVec,
    candidatesJson,
    topK,
    applySourceWeights,
  ) as CoreRankedResult[];

  // Map core results back to RerankerResult, restoring the optional
  // metadata (importance/createdAt/source) so callers that still read those
  // fields keep working.
  const candidateMap = new Map<string, RerankerCandidate>();
  for (const c of candidates) candidateMap.set(c.id, c);

  return ranked.map((r) => {
    const orig = candidateMap.get(r.id);
    return {
      id: r.id,
      text: r.text,
      embedding: orig?.embedding,
      importance: orig?.importance,
      createdAt: orig?.createdAt,
      source: orig?.source,
      rrfScore: r.score,
      cosineSimilarity: r.cosine_score,
      sourceWeight: applySourceWeights ? (r.source_weight ?? 1.0) : undefined,
    };
  });
}

File: python/src/totalreclaw/reranker.py
"""TotalReclaw Hermes (Python client) -- reranker wrapper.

As of rc.22 the Python client no longer ships its own BM25 / RRF / source-weight
implementation. All ranking decisions are delegated to the canonical Rust
core via `totalreclaw_core.rerank_with_config`, `totalreclaw_core.cosine_similarity`,
`totalreclaw_core.source_weight`, and `totalreclaw_core.legacy_claim_fallback_weight`.
This guarantees plugin / Hermes / MCP runtimes share one ranker source of truth
and cannot drift again (rc.18 cosine-gate divergence root cause).

The previous Python pipeline added importance, recency, and MMR signals on
top of BM25 + cosine. Those signals are dropped here: core's intent-weighted
RRF + source-weighted final score is the canonical Pipeline G + Tier 1 mix
that benchmark E13 calibrated. The extra Python-side passes were not part
of the validated baseline and were a source of cross-client divergence.

Public surface kept for callers:
  - ``rerank(query, query_embedding, candidates, top_k, apply_source_weights)``
    -- called by ``operations.py``
  - ``cosine_similarity(a, b)`` -- called by ``agent/lifecycle.py``
  - ``source_weight(source)`` / ``LEGACY_CLAIM_FALLBACK_WEIGHT``
  - Dataclasses: ``RerankerCandidate``, ``RerankerResult`` -- ``RerankerResult``
    is re-exported by ``client.py`` and read by ``agent/contradiction.py``

Core owns intent-weighting now (per-candidate BM25/cosine ratio), so the old
client-side ``detect_query_intent`` / ``INTENT_WEIGHTS`` / ``RankingWeights``
compat surface has been removed -- no internal caller used it.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Optional

import totalreclaw_core


# ---------------------------------------------------------------------------
# Cosine Similarity (delegated to core)
# ---------------------------------------------------------------------------

def cosine_similarity(a: list[float], b: list[float]) -> float:
    """Compute cosine similarity by delegating to ``totalreclaw_core``.

    Returns 0 if either vector is empty. Equal-length vectors are required;
    we truncate to ``min(len(a), len(b))`` when lengths differ to match the
    historical Python behaviour.
    """
    if not a or not b:
        return 0.0
    length = min(len(a), len(b))
    a_trunc = [float(x) for x in a[:length]]
    b_trunc = [float(x) for x in b[:length]]
    return totalreclaw_core.cosine_similarity(a_trunc, b_trunc)


# ---------------------------------------------------------------------------
# Source weight (delegated to core)
# ---------------------------------------------------------------------------

LEGACY_CLAIM_FALLBACK_WEIGHT: float = float(totalreclaw_core.legacy_claim_fallback_weight())


def source_weight(source: Optional[str]) -> float:
    """Return the v1 source weight from core.

    ``None`` (or empty string) returns the legacy-claim fallback weight.
    Unknown source strings are routed through core's ``MemorySource::from_str_lossy``
    which lands them at ``user-inferred`` (0.9) -- the canonical safe default.
    """
    if not source:
        return LEGACY_CLAIM_FALLBACK_WEIGHT
    return float(totalreclaw_core.source_weight(source))


# ---------------------------------------------------------------------------
# Data classes (cross-runtime stable surface)
# ---------------------------------------------------------------------------

@dataclass
class RerankerCandidate:
    id: str
    text: str
    embedding: Optional[list[float]] = None
    importance: Optional[float] = None
    created_at: Optional[float] = None  # Unix timestamp (seconds)
    category: str = "fact"
    #: v1 provenance source ("user", "user-inferred", "assistant",
    #: "external", "derived"). When None, the candidate is treated as a
    #: pre-v1 legacy blob and receives ``LEGACY_CLAIM_FALLBACK_WEIGHT``.
    source: Optional[str] = None
    #: Filtered claim-metadata passthrough (#425) — e.g. import_source /
    #: session_id / subtype. Never sent to core; rejoined by id after
    #: ``rerank_with_config`` so read surfaces can expose provenance.
    metadata: Optional[dict] = None


@dataclass
class RerankerResult:
    """Result from the reranker -- mirrors RerankerCandidate fields plus scores."""
    id: str
    text: str
    embedding: Optional[list[float]] = None
    importance: Optional[float] = None
    created_at: Optional[float] = None
    category: str = "fact"
    rrf_score: float = 0.0
    cosine_sim: Optional[float] = None
    source: Optional[str] = None
    source_weight: Optional[float] = None
    metadata: Optional[dict] = None


# ---------------------------------------------------------------------------
# Re-ranker (delegates to core::reranker)
# ---------------------------------------------------------------------------

def rerank(
    query: str,
    query_embedding: Optional[list[float]],
    candidates: list[RerankerCandidate],
    top_k: int = 8,
    apply_source_weights: bool = False,
) -> list[RerankerResult]:
    """Re-rank candidates by routing through ``totalreclaw_core.rerank_with_config``.

    Core handles intent-weighting itself based on the per-candidate cosine
    score. Importance / recency / MMR signals are no longer applied
    client-side; if you need those, contribute them to core::reranker first.
    """
    if not candidates:
        return []

    # Build the core::Candidate JSON shape:
    #   { id, text, embedding, timestamp, source? }
    # ``embedding`` defaults to [] for candidates without an embedding;
    # core's cosine_similarity_f32 returns 0 for that case.
    core_candidates: list[dict] = []
    for c in candidates:
        core_obj: dict = {
            "id": c.id,
            "text": c.text,
            "embedding": list(c.embedding) if c.embedding else [],
            "timestamp": str(c.created_at) if c.created_at is not None else "",
        }
        if c.source:
            core_obj["source"] = c.source
        core_candidates.append(core_obj)

    candidates_json = json.dumps(core_candidates)
    query_vec = [float(x) for x in (query_embedding or [])]

    raw = totalreclaw_core.rerank_with_config(
        query,
        query_vec,
        candidates_json,
        top_k,
        apply_source_weights,
    )
    # rerank_with_config returns a JSON string of RankedResult objects.
    ranked: list[dict] = json.loads(raw)

    # Restore the optional metadata (importance / created_at / category /
    # embedding / source) so consumers like agent/contradiction.py keep
    # working without changes.
    by_id = {c.id: c for c in candidates}
    out: list[RerankerResult] = []
    for r in ranked:
        orig = by_id.get(r["id"])
        out.append(
            RerankerResult(
                id=r["id"],
                text=r["text"],
                embedding=orig.embedding if orig else None,
                importance=orig.importance if orig else None,
                created_at=orig.created_at if orig else None,
                category=orig.category if orig else "fact",
                rrf_score=float(r.get("score", 0.0)),
                cosine_sim=float(r.get("cosine_score", 0.0)),
                source=orig.source if orig else None,
                source_weight=float(r["source_weight"])
                if apply_source_weights and "source_weight" in r
                else (1.0 if apply_source_weights else None),
                metadata=orig.metadata if orig else None,
            )
        )

    return out
